x/gov/simulation/operations.go

```diff
@@ -0,0 +1,312 @@
+		account := ak.GetAccount(ctx, simAccount.Address)
+		coins := account.SpendableCoins(ctx.BlockTime())
+
+		var fees sdk.Coins
+		coins, hasNeg := coins.SafeSub(deposit)
+		if !hasNeg {
+			fees, err = simulation.RandomFees(r, ctx, coins)
+			if err != nil {
+				return simulation.NoOpMsg(types.ModuleName), nil, err
+			}
+		}
+
+		tx := helpers.GenTx(
+			[]sdk.Msg{msg},
+			fees,
+			chainID,
+			[]uint64{account.GetAccountNumber()},
+			[]uint64{account.GetSequence()},
+			[]crypto.PrivKey{simAccount.PrivKey}...,
+		)
+
+		res := app.Deliver(tx)
+		if !res.IsOK() {
+			return simulation.NoOpMsg(types.ModuleName), nil, errors.New(res.Log)
+		}
+
+		return simulation.NewOperationMsg(msg, true, ""), nil, nil
+	}
+}
+
+// SimulateMsgVote generates a MsgVote with random values.
+func SimulateMsgVote(ak types.AccountKeeper, k keeper.Keeper) simulation.Operation {
+	return operationSimulateMsgVote(ak, k, simulation.Account{}, -1)
+}
+
+func operationSimulateMsgVote(ak types.AccountKeeper, k keeper.Keeper,
+	simAccount simulation.Account, proposalIDInt int64) simulation.Operation {
+	return func(r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, accs []simulation.Account,
+		chainID string) (simulation.OperationMsg, []simulation.FutureOperation, error) {
+		if simAccount.Equals(simulation.Account{}) {
+			simAccount, _ = simulation.RandomAcc(r, accs)
+		}
+
+		var proposalID uint64
+
+		switch {
+		case proposalIDInt < 0:
+			var ok bool
+			proposalID, ok = randomProposalID(r, k, ctx, types.StatusVotingPeriod)
+			if !ok {
+				return simulation.NoOpMsg(types.ModuleName), nil, nil
+			}
+		default:
+			proposalID = uint64(proposalIDInt)
+		}
+
+		option := randomVotingOption(r)
+
+		msg := types.NewMsgVote(simAccount.Address, proposalID, option)
+
+		account := ak.GetAccount(ctx, simAccount.Address)
+		fees, err := simulation.RandomFees(r, ctx, account.SpendableCoins(ctx.BlockTime()))
+		if err != nil {
+			return simulation.NoOpMsg(types.ModuleName), nil, err
+		}
+
+		tx := helpers.GenTx(
+			[]sdk.Msg{msg},
+			fees,
+			chainID,
+			[]uint64{account.GetAccountNumber()},
+			[]uint64{account.GetSequence()},
+			[]crypto.PrivKey{simAccount.PrivKey}...,
+		)
+
+		res := app.Deliver(tx)
+		if !res.IsOK() {
+			return simulation.NoOpMsg(types.ModuleName), nil, errors.New(res.Log)
+		}
+
+		return simulation.NewOperationMsg(msg, true, ""), nil, nil
+	}
+}
+
+// Pick a random deposit
+func randomDeposit(r *rand.Rand, ctx sdk.Context,
+	ak types.AccountKeeper, k keeper.Keeper, addr sdk.AccAddress,
+) (deposit sdk.Coins, skip bool, err error) {
+	account := ak.GetAccount(ctx, addr)
+	coins := account.SpendableCoins(ctx.BlockHeader().Time)
+	if coins.Empty() {
+		return nil, true, nil // skip
+	}
+
+	minDeposit := k.GetDepositParams(ctx).MinDeposit
+	denomIndex := r.Intn(len(minDeposit))
+	denom := minDeposit[denomIndex].Denom
+
+	depositCoins := coins.AmountOf(denom)
+	if depositCoins.IsZero() {
+		return nil, true, nil
+	}
+
+	maxAmt := depositCoins
```

Comment:
Does this make sense? If we choose a random deposit amt > min deposit, then we use the min deposit. Why?

Shouldn't it rather be:

```go
if depositAmt.LT(minDeposit[denomIndex]) {
    depositAmt = minDeposit[denomIndex]
}
```
```

Reply:
My reasoning was that the proposer deposits an amount which calculated randomly between `(0, min(minDepositAmt, balance)]`. 

In the case you suggested, the deposit amount would always be ≥ than the min deposit, which is not what we see in the Hub or other zones as users "crowdfund" the proposal by sending deposit txs to it when the proposer didn't deposit the minimum required.

Reply:
the `[denomIndex]` is correct tho, good catch

Reply:
Ulresolving; I don't see the godoc state the logic you've described.